utils/signature/verify.go

```diff
@@ -7,58 +7,121 @@ SPDX-License-Identifier: Apache-2.0
 	res := &NsVerifier{
-		Scheme:    strings.ToUpper(scheme),
-		PublicKey: key,
+		NamespacePolicy: p,
 	}
 	var err error
-	switch res.Scheme {
-	case NoScheme, "":
-		res.verifier = nil
-	case Ecdsa:
-		res.verifier, err = NewEcdsaVerifier(key)
-	case Bls:
-		res.verifier, err = NewBLSVerifier(key)
-	case Eddsa:
-		res.verifier = &EdDSAVerifier{PublicKey: key}
+
+	switch r := p.GetRule().(type) {
+	case *protoblocktx.NamespacePolicy_ThresholdRule:
+		policy := r.ThresholdRule
+
+		switch policy.Scheme {
+		case NoScheme, "":
+			res.verifier = nil
+		case Ecdsa:
+			res.verifier, err = newEcdsaVerifier(policy.PublicKey)
+		case Bls:
+			res.verifier, err = newBLSVerifier(policy.PublicKey)
+		case Eddsa:
+			res.verifier = &edDSAVerifier{PublicKey: policy.PublicKey}
+		default:
+			return nil, errors.Newf("scheme '%v' not supported", policy.Scheme)
+		}
+	case *protoblocktx.NamespacePolicy_SignatureRule:
+		pp := cauthdsl.NewPolicyProvider(idDeserializer)
+		res.verifier, _, err = pp.NewPolicy(r.SignatureRule)
 	default:
-		return nil, errors.Newf("scheme '%v' not supported", scheme)
+		return nil, errors.Newf("policy rule '%v' not supported", p.GetRule())
 	}
 	return res, err
 }
 
 // VerifyNs verifies a transaction's namespace signature.
 func (v *NsVerifier) VerifyNs(txID string, tx *protoblocktx.Tx, nsIndex int) error {
-	if nsIndex < 0 || nsIndex >= len(tx.Namespaces) || nsIndex >= len(tx.Signatures) {
+	if nsIndex < 0 || nsIndex >= len(tx.Namespaces) || nsIndex >= len(tx.Endorsements) {
 		return errors.New("namespace index out of range")
 	}
+
 	if v.verifier == nil {
 		return nil
 	}
-	digest, err := DigestTxNamespace(txID, tx.Namespaces[nsIndex])
+
+	data, err := ASN1MarshalTxNamespace(txID, tx.Namespaces[nsIndex])
 	if err != nil {
 		return err
 	}
-	return v.verifier.Verify(digest, tx.Signatures[nsIndex])
+
+	switch v.NamespacePolicy.GetRule().(type) {
+	case *protoblocktx.NamespacePolicy_ThresholdRule:
+		return v.verifier.EvaluateSignedData([]*protoutil.SignedData{
+			{
+				Data:      data,
+				Signature: tx.Endorsements[nsIndex].EndorsementsWithIdentity[0].Endorsement,
+			},
+		})
+	case *protoblocktx.NamespacePolicy_SignatureRule:
+		signedData := make([]*protoutil.SignedData, len(tx.Endorsements[nsIndex].EndorsementsWithIdentity))
+		for i, s := range tx.Endorsements[nsIndex].EndorsementsWithIdentity {
+			// NOTE: CertificateID is not supported as MSP does not have the supported for pre-stored certificates yet.
+			cert := s.Identity.GetCertificate()
+			if cert == nil {
+				return errors.New("An empty certificate is provided for the identity")
+			}
+			idBytes, err := msp.NewSerializedIdentity(s.Identity.MspId, cert)
+			if err != nil {
+				return err
+			}
+
+			signedData[i] = &protoutil.SignedData{
+				Data:      data,
+				Identity:  idBytes,
+				Signature: s.Endorsement,
+			}
+		}
+		return v.verifier.EvaluateSignedData(signedData)
```

Comment:
nit: `signedData` can be pre-declared outside the switch and have a single return statement for both cases.